firebase-perf/src/test/java/com/google/firebase/perf/application/FrameMetricsRecorderTest.java

```diff
@@ -0,0 +1,261 @@
+    ArgumentCaptor<Activity> activityCaptor = ArgumentCaptor.forClass(Activity.class);
+    recorder.start();
+    recorder.stop();
+    recorder.start();
+    recorder.stop();
+    verify(fma, times(2)).add(activityCaptor.capture());
+    verify(fma, times(2)).remove(activityCaptor.capture());
+    List<Activity> activities = activityCaptor.getAllValues();
+    assertThat(activities.stream().allMatch(a -> a == activity)).isTrue();
+  }
+
+  @Test
+  public void start_whileAlreadyStarted_doesNotCallFMATwice() {
+    recorder.start();
+    recorder.start();
+    verify(fma, times(1)).add(any());
+  }
+
+  @Test
+  public void start_afterPreviousEnded_doesCallFMASuccessfully() {
+    recorder.start();
+    recorder.stop();
+    recorder.start();
+    verify(fma, times(2)).add(any());
+  }
+
+  @Test
+  public void stop_whileNotStarted_returnsEmptyResult() {
+    Optional<PerfFrameMetrics> result;
+    result = recorder.stop();
+    assertThat(result.isAvailable()).isFalse();
+  }
+
+  @Test
+  public void stop_calledTwice_returnsEmptyResult() {
+    Optional<PerfFrameMetrics> result;
+    recorder.start();
+    recorder.stop();
+    result = recorder.stop();
+    assertThat(result.isAvailable()).isFalse();
+  }
+
+  @Test
+  public void startAndStop_calledInCorrectOrder_returnsValidResult() {
+    Optional<PerfFrameMetrics> result;
+    recorder.start();
+    result = recorder.stop();
+    assertThat(result.isAvailable()).isTrue();
+    assertThat(result.get().getTotalFrames()).isEqualTo(3);
+    assertThat(result.get().getSlowFrames()).isEqualTo(2);
+    assertThat(result.get().getFrozenFrames()).isEqualTo(1);
+  }
+
+  @Test
+  public void startAndStopSubTrace_activityRecordingNeverStarted_returnsEmptyResult() {
+    Fragment fragment = new Fragment();
+    Optional<PerfFrameMetrics> result;
+
+    stubFrameMetricsAggregatorData(fma, frameTimes1);
+    recorder.startSubTrace(fragment);
+    stubFrameMetricsAggregatorData(fma, frameTimes2);
+    result = recorder.stopSubTrace(fragment);
+    assertThat(result.isAvailable()).isFalse();
+  }
+
+  @Test
+  public void startAndStopSubTrace_activityRecordingHasEnded_returnsEmptyResult() {
+    Fragment fragment = new Fragment();
+    Optional<PerfFrameMetrics> result;
+
+    recorder.start();
+    recorder.stop();
+    stubFrameMetricsAggregatorData(fma, frameTimes1);
+    recorder.startSubTrace(fragment);
+    stubFrameMetricsAggregatorData(fma, frameTimes2);
+    result = recorder.stopSubTrace(fragment);
+    assertThat(result.isAvailable()).isFalse();
+  }
+
+  @Test
+  public void startAndStopSubTrace_whenCalledTwice_ignoresSecondCall() {
+    Fragment fragment = new Fragment();
+    Optional<PerfFrameMetrics> result1;
+    Optional<PerfFrameMetrics> result2;
+
+    // Happens only when we hook on to the same instance of fragment twice. Very unlikely.
+    recorder.start();
+    stubFrameMetricsAggregatorData(fma, new int[][] {{1, 1}, {17, 1}, {800, 1}}); // 3, 2, 1
```

Comment:
I think it would be helpful to have comments on each of these explaining whether we expect them to be taken into account or not and why